File: session/01_concurrency_and_parallelism/NationalStockService.java
import java.io.*;
import java.net.*;

import org.json.JSONObject;

public class NationalStockService {
  private final String urlTemplate = "http://national-stock-service.herokuapp.com:5000/stocks/%s";

  public double getPrice(final String ticker) throws Exception {
    System.out.println(Thread.currentThread() + " Getting Price for => " + ticker);
    final URL url = new URL(String.format(urlTemplate, ticker));
    final String data = getData(url.openConnection());
    final JSONObject stockDetails = new JSONObject(data);
    double price = stockDetails.getDouble("price");
    System.out.println(Thread.currentThread() + " Returning Price for => " + ticker + " price = " + price);
    return price;
  }

  private String getData(final URLConnection connection) throws IOException {
    InputStream is = connection.getInputStream();
    InputStreamReader isr = new InputStreamReader(is);
    BufferedReader reader = new BufferedReader(isr);
    String line = null;
    StringBuilder data = new StringBuilder();
    while ((line = reader.readLine()) != null) {
      data.append(line);
    }
    return data.toString();
  }
}

File: melodies/concurrency_and_parallelism/parallelism/splitting_io_task/scala/NationalStockService.scala
package tsys.stocks

import scala.util.{Try, Success, Failure}

class NationalStockService {
  private val urlTemplate = "https://national-stock-service.herokuapp.com/stocks/"
  
  def getPrice(ticker: String): Try[Double] = {
    val url = s"$urlTemplate$ticker"
    println(s"Fetching $ticker price using ${Thread.currentThread()}")
    Try {
      io.Source.fromURL(url).mkString
    }.flatMap(data => {
      import scala.util.parsing.json._
      val json: Option[Any] = JSON.parseFull(data)
      val price: Option[Double] = json.map(response => {
        val values = response.asInstanceOf[Map[String, Any]]
        values("price").asInstanceOf[Double]
      })
      Try { price.get }
    })
  }
}